### PR TITLE
Flambda2 Validator

### PR DESCRIPTION
**Pull request by euisuny**
This draft PR is a proof-of-concept version of the Flambda2 validator.

## Structure

The validator resides in `middle_end/flambda2/validate/`.
- `flambda2_core.ml` : the Flambda2 Core language
There are several differences between the `Flambda2` IR and `Flambda2 Core`, and the most important are the following:
1. Flambda2 allows for full subexpressions in argument position, so that there can be more aggressive inlining in the normalization process. Since allowing for full subexpressions in the argument position and full inlining is not always sound (especially under the presence of effects), the normalization for `Flambda2 Core` needs to be conservative under certain conditions. This draft PR *does not* support sound reduction for `Flambda2 Core` with effects and mutable state.
2. Set of closures binding
In `Flambda2`, there are two ways that a binding for a set of closures can be declared: 
    1. As a variable list, where the list of variables correspond to each function stored in the closure.
     `let f = closure f_id @f and g = closure g_id @g` corresponds to binding two variables `f` and `g` each corresponding to`@f` and `@g` are globally scoped function slots which map to code ids `f_id` and `g_id`.
    2. As a static list of symbols, where the list of symbols correspond to each function stored in the closure.
      `let $f = closure f_id @f and $g = closure g_id @g`.

    In `Flambda2 Core`, there is only one way where a set of closures can be declared (note that we can afford to be more lenient with the representation because variables, in general, do not have to be register-sized in `Flambda2 Core`).
    `let Φ = { closure f_id @f;  closure g_id @g}` where `Φ` is a variable corresponding to the whole set of closures. Each of the function slots can be referred to by using a "slot expression", which is a tuple referring to the set of closures and the function slot. to referred to the function slot that stores `f_id`, we write `slot (Φ, @f)`.

    Because code expressions are fully substituted into sets of closures, this change in set of closures binding causes other differences in the `Flambda2 Core` language which we elide here.

- `translate.ml` : Translation from `Flambda2` to `Flambda2 Core`
- `equiv.ml` : An alpha-equivalence checker for `Flambda2 Core`, largely inspired by @lukemaurer 's alpha equivalence checker for `Flambda2`. Several things to note for the `Flambda2` (and `Flambda2 Core`) binding behavior.
    1. The static symbols are `dominator scoped`. Dominator scoping assumes that (akin to Single Static Assignment), variables are given a static, unique assignment at each invocation. The scoping is determined by the control flow, instead of lexical scope.
    2. The function slots and value slots in sets of closures do not have a explicit binding site; they have global scope.
    3. The `done` continuation (top-level continuation) is a global identifier.
- `bound_identifiers/` : This directory contains the binders in `Flambda2 Core` that differ from `Flambda2`.
- `normalize.ml` : The normalization scheme for Flambda2. The normalization reduces as eagerly as possible, and also does reduction under lambda bindings.
- `eval_prim.ml` : Evaluation of primitive operations. Block-based primitives (`Block_load`, `Make_block`) are implemented, and the arithmetic primitives are not yet implemented.

## Proof of concept
The validator is currently able to validate simple programs that have recursion and basic data types.
It is able to validate the (among other `.fl` files in the test suite described in a below section) `Flambda2` program which is generated from the following `ml` snippet. This corresponds to `test13.ml`, taken from an already-existing test in `flambda2/tests/mlexamples/`.
```
external opaque : 'a -> 'a = "%opaque"

external ignore : 'a -> unit = "%ignore"

external ( + ) : int -> int -> int = "%addint"

type +'a node =
  | Nil
  | Cons of 'a * 'a t

and 'a t = unit -> 'a node

let[@inline always] bar i map_foo =
  ();
  fun () ->
    let j = i + 3 in
    ignore (opaque j);
    (map_foo [@inlined never]) (fun x -> x) (fun () -> Nil) ()

let rec map_foo f seq () =
  match seq () with
  | Nil -> Nil
  | Cons (x, next) ->
    let (g : unit -> 'a node) =
      if opaque false then bar 10 map_foo else bar 20 map_foo
    in
    let g_result = (g [@inlined always]) () in
    ignore (opaque g_result);
    Cons (f x, fun () -> Nil)
```

## Using the validator in compilation
In order to run the validator in the simplifier pass, run the compiler with the `-validate` flag.

## Testing the validator
The test suite for the validator is in `test-validate` and the `test-validate/tests` contains `*.fl` test cases that can be used to test the validator.

To add a test in the suite, add a new `.fl` file under `test-validate/tests`, and run `inc.sh` (which generates the appropriate dune rules to run the test suite).

#### Verbose testing
Run `make hacking-runtest` and add `normalize_term "<insert fl file>"` to the main function in `test-validate/flambda2_validation.ml`. The verbose testing will print the input and output of the simplifer for `Flambda2`, and also the results post-translation and normalization.

### CODE REVIEW

**Comment by Gbury** [comment]
Maybe that's a weird question, but what do you mean exactly by "validate" ? Does it mean checking some invariants, and if so, which invariants ?

**Comment by euisuny** [comment]
It's a checker for semantic equivalence between terms that are input and output by the Flambda2 simplifier.
Given the simplifying pass (`Simplify.run`) this validator does the following:


      [Flambda_unit.t]   ------- Simplify.run --------> [simplify_result]
              |                                                 |
              |                                                 |
          translate                                         translate
    (flambda_expr_to_core)                          (simplify_result_to_core)
              |                                                 |
              |                                                 |
              ↓                                                 ↓
          [core_exp]                                        [core_exp]
              |                                                 |
              |                                                 |
          β - reduce                                        β - reduce
          (normalize)                                       (normalize)
              |                                                 |
              |                                                 |
              ↓                                                 ↓
        [core_exp]              ≅[validate]                [core_exp]

File: middle_end/flambda2/validate/equiv.ml [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+  let add_symbol t symbol1 symbol2 =
+    t.symbols <- Symbol.Map.add symbol1 symbol2 t.symbols
+
+  let add_code_id t code_id1 code_id2 =
+    t.code_ids <- Code_id.Map.add code_id1 code_id2 t.code_ids
+
+  let add_function_slot t function_slot1 function_slot2 =
+    t.function_slots
+      <- Function_slot.Map.add function_slot1 function_slot2 t.function_slots;
+    t.function_slots
```
Comment (xclerc) [inline middle_end/flambda2/validate/equiv.ml:43]:
```suggestion
    t.function_slots_rev
```

File: middle_end/flambda2/validate/equiv.ml [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+      <- Function_slot.Map.add function_slot2 function_slot1
+           t.function_slots_rev
+
+  let add_value_slot t value_slot1 value_slot2 =
+    t.value_slots <- Value_slot.Map.add value_slot1 value_slot2 t.value_slots;
+    t.value_slots
```
Comment (xclerc) [inline middle_end/flambda2/validate/equiv.ml:49]:
```suggestion
    t.value_slots_rev
```

File: middle_end/flambda2/validate/equiv.ml [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+  Env.find_code_id env code_id |> Option.value ~default:code_id
+
+let subst_function_expr env (fn_expr : function_expr) =
+  match fn_expr with
+  | Id id -> Id (subst_code_id env id)
+  | Exp _ -> fn_expr
+
+let _subst_code_id (env : Env.t) code_id =
+  Env.find_code_id env code_id |> Option.value ~default:code_id
+
```
Comment (xclerc) [inline middle_end/flambda2/validate/equiv.ml:101]:
```suggestion
```
(duplicate of line 91)

**Comment by ccasin** [comment]
[These](https://github.com/ocaml-flambda/flambda-backend/pull/1200/commits/f85677b78d3ff00242c52cfa99f6fa8a9dc53cea) [two](https://github.com/ocaml-flambda/flambda-backend/pull/1200/commits/f85677b78d3ff00242c52cfa99f6fa8a9dc53cea) commits have snuck into your PR (I think this may be because there was a force push to the main branch).  Could you rebase to remove them?

**Comment by euisuny** [comment]
Leaving a note here for keeping track of offline review: [c70e5a5](https://github.com/ocaml-flambda/flambda-backend/pull/1181/commits/c70e5a5b4a44cc4b5ab25f865471aee625a698ec) addresses the comments and suggestions from offline review with @ccasin

**Comment by mshinwell** [comment]
@ccasin what is the plan for this work?

**Comment by ccasin** [comment]
@mshinwell Work continues on this, slowly.  The PR mainly existed for us to review an intermediate state, so I'll close it for now with a plan to re-open when we're done.

**Pull request closed without merging**